File: 23_perfect_square/23_perfect_square.py
# python program to check whether the number is a perfect square or not
# a number is a perfect square if its root is an integer value.

# method 1 - checking equality of square
n=int(input("Enter a number :"))
m=int(n**(0.5))
if m**2 == n:
    print(f"{n} is a perfect square")
else:
    print(f"{n} is not a perfect square")

# method 2 - checking equality of root
n=int(input("Enter a number :"))
m=n**(0.5)
l=int(m)
if m == l:
    print(f"{n} perfect square")
else:
    print(f"{n} not a Perfect Square")

# method 3 - module based approach

def is_perfect_square(num):
    square_root = int(num ** (0.5))
    if (square_root * square_root):
        return True
    return False

try:
    while(True):

        num = int(input("Enter positive integer value : "))
        limit  = 1000
        if (num <= 0):
            print("Enter positiveinteger value")
        elif (num > limit):
            print("Input too large")
        else:
            result = is_perfect_square(num)
            if(result):
                print(f"{num} is a perfect square")
            else:
                print(f"{num} is not a perfect square")

        choice = input("Do you want to continue the program (y/n) : ")
        if(choice == "n"):
            print("Program finished!")
            break

except ValueError as e:
    print("Invalid Input : {e}")

except ZeroDivisionError as e:
    print(f"Zero division exception : {e}")

except Exception as e:
    print("Exception caught : {e}")

    

File: 23_perfect_square/23_perfect_square.dart
import 'dart:io';
import 'dart:math';

// program to check whether a number is a perfect square

bool isPerfectSquare(int num) {
  int squareRoot = sqrt(num).toInt();
  if (squareRoot * squareRoot == num) {
    return true;
  }
  return false;
}

void main() {
  try {
    while (true) {
      stdout.write("Enter a positive integer value : ");
      int num = int.parse(stdin.readLineSync()!);

      int maxLimit = 1000;

      if (num > maxLimit) {
        print("Enter a lesser value than 10000");
      } else if (num <= 0) {
        print("Enter a valid integer value : ");
      } else {
        bool result = isPerfectSquare(num);
        if (result) {
          print("$num is a perfect square");
        } else {
          print("$num is not a perfect square");
        }
      }

      stdout.write("Do you want to continue the program (y/n) : ");
      String choice = stdin.readLineSync()!.toLowerCase();
      if (choice == 'n') {
        print("Program finished!");
        break;
      }
    }
  } on FormatException catch (e) {
    print("Invalid input : $e");
  } on UnsupportedError catch (e) {
    print("Divide by zero exception : $e");
  } on Exception catch (e) {
    print("Exception caught : $e");
  }
}
